tests/src/tests.rs

```diff
@@ -543,92 +555,158 @@ fn test_part(
     }
 
     // Map errors to range and message format, discard traces and errors from
-    // other files.
-    let mut errors: Vec<_> = errors
+    // other files. Collect hints.
+    //
+    // This has one caveat: due to the format of the expected hints, we can not verify if a hint belongs
+    // to a error or not. That should be irrelevant however, as the line of the hint is still verified.
+    let actual_errors_and_hints: HashSet<UserOutput> = errors
         .into_iter()
         .filter(|error| error.span.source() == id)
-        .map(|error| (error.range(world), error.message.replace('\\', "/")))
+        .flat_map(|error| {
+            let output_error =
+                UserOutput::Error(error.range(world), error.message.replace('\\', "/"));
+            let hints = error
+                .hints
+                .iter()
+                .filter(|_| validate_hints) // No unexpected hints should be verified if disabled.
+                .map(|hint| UserOutput::Hint(error.range(world), hint.to_string()));
+            iter::once(output_error).chain(hints).collect::<Vec<_>>()
+        })
         .collect();
 
-    errors.sort_by_key(|error| error.0.start);
-    ref_errors.sort_by_key(|error| error.0.start);
+    // Basically symmetric_difference, but we need to know where an item is coming from.
+    let mut unexpected_outputs = actual_errors_and_hints
+        .difference(&metadata.invariants)
+        .collect::<Vec<_>>();
+    let mut missing_outputs = metadata
+        .invariants
+        .difference(&actual_errors_and_hints)
+        .collect::<Vec<_>>();
+
+    unexpected_outputs.sort_by_key(|&o| o.start());
+    missing_outputs.sort_by_key(|&o| o.start());
 
-    if errors != ref_errors {
+    // This prints all unexpected emits first, then all missing emits.
```

Comment:
Remark to the reviewer: This comment is more of a remark to the reviewer, I'll happily remove it if preferred.